File: bindings/rust/src/graph/queries.rs
//! Query operations for Graph.

use crate::utils::{escape_string, rel_type_pattern};
use crate::{CypherResult, Result, Value};
use super::{Graph, GraphStats};

impl Graph {
    /// Get the degree (number of connections) of a node.
    pub fn node_degree(&self, node_id: &str) -> Result<i64> {
        let query = format!(
            "MATCH (n {{id: '{}'}})-[r]-() RETURN count(r) AS degree",
            escape_string(node_id)
        );
        let result = self.connection().cypher(&query)?;
        if result.is_empty() {
            return Ok(0);
        }
        Ok(result[0].get("degree").unwrap_or(0))
    }

    /// Get all neighboring nodes (connected via any edge direction).
    pub fn get_neighbors(&self, node_id: &str) -> Result<Vec<Value>> {
        let query = format!(
            "MATCH (n {{id: '{}'}})-[]-(m) RETURN DISTINCT m",
            escape_string(node_id)
        );
        let result = self.connection().cypher(&query)?;
        let mut neighbors = Vec::new();
        for row in result.iter() {
            if let Some(m) = row.get_value("m") {
                neighbors.push(m.clone());
            }
        }
        Ok(neighbors)
    }

    /// Get graph statistics (node and edge counts).
    pub fn stats(&self) -> Result<GraphStats> {
        let nodes_result = self.connection().cypher("MATCH (n) RETURN count(n) AS cnt")?;
        let edges_result = self.connection().cypher("MATCH ()-[r]->() RETURN count(r) AS cnt")?;

        let nodes = if nodes_result.is_empty() {
            0
        } else {
            nodes_result[0].get("cnt").unwrap_or(0)
        };

        let edges = if edges_result.is_empty() {
            0
        } else {
            edges_result[0].get("cnt").unwrap_or(0)
        };

        Ok(GraphStats { node_count: nodes, edge_count: edges })
    }

    /// Get all outgoing edges from a node.
    ///
    /// Returns a [`CypherResult`] with columns: `source`, `target`, `r`.
    pub fn get_edges_from(&self, node_id: &str) -> Result<CypherResult> {
        let query = format!(
            "MATCH (a {{id: '{}'}})-[r]->(b) RETURN a.id AS source, b.id AS target, r",
            escape_string(node_id)
        );
        self.connection().cypher(&query)
    }

    /// Get all incoming edges to a node.
    ///
    /// Returns a [`CypherResult`] with columns: `source`, `target`, `r`.
    pub fn get_edges_to(&self, node_id: &str) -> Result<CypherResult> {
        let query = format!(
            "MATCH (a)-[r]->(b {{id: '{}'}}) RETURN a.id AS source, b.id AS target, r",
            escape_string(node_id)
        );
        self.connection().cypher(&query)
    }

    /// Get outgoing edges of a specific type from a node.
    ///
    /// Returns a [`CypherResult`] with columns: `source`, `target`, `r`.
    pub fn get_edges_by_type(&self, node_id: &str, rel_type: &str) -> Result<CypherResult> {
        let rel = rel_type_pattern(Some(rel_type));
        let query = format!(
            "MATCH (a {{id: '{}'}})-[r{}]->(b) RETURN a.id AS source, b.id AS target, r",
            escape_string(node_id),
            rel
        );
        self.connection().cypher(&query)
    }

    /// Get all edges (both directions) connected to a node.
    ///
    /// Returns a [`CypherResult`] with columns: `source`, `target`, `r`.
    pub fn get_node_edges(&self, node_id: &str) -> Result<CypherResult> {
        let query = format!(
            "MATCH (n {{id: '{}'}})-[r]-(m) RETURN n.id AS source, m.id AS target, r",
            escape_string(node_id)
        );
        self.connection().cypher(&query)
    }
}

File: bindings/python/src/graphqlite/graph/queries.py
"""Query operations mixin for Graph class."""

from typing import Optional

from ._base import BaseMixin


class QueriesMixin(BaseMixin):
    """Mixin providing graph query operations."""

    def node_degree(self, node_id: str) -> int:
        """
        Get the degree (number of connections) of a node.

        Counts both incoming and outgoing edges.

        Args:
            node_id: The node's id property value

        Returns:
            Number of edges connected to the node
        """
        result = self._conn.cypher(
            "MATCH (n {id: $id})-[r]-() "
            "RETURN count(r) AS degree",
            params={"id": node_id},
        )
        if len(result) == 0:
            return 0
        deg = result[0].get("degree", 0)
        return int(deg) if deg else 0

    def get_neighbors(self, node_id: str) -> list[dict]:
        """
        Get all neighboring nodes.

        Returns nodes connected by edges in either direction.

        Args:
            node_id: The node's id property value

        Returns:
            List of neighbor node dicts
        """
        result = self._conn.cypher(
            "MATCH (n {id: $id})-[]-(m) "
            "RETURN DISTINCT m",
            params={"id": node_id},
        )
        return [row.get("m") for row in result if row.get("m")]

    def get_node_edges(self, node_id: str) -> list[tuple[str, str, dict]]:
        """
        Get all edges connected to a node.

        Args:
            node_id: The node's id property value

        Returns:
            List of (source_id, target_id, properties) tuples
        """
        result = self._conn.cypher(
            "MATCH (n {id: $id})-[r]-(m) "
            "RETURN n.id AS source, m.id AS target, r",
            params={"id": node_id},
        )
        return [(row["source"], row["target"], row.get("r", {})) for row in result]

    def get_edges_from(self, node_id: str) -> list[dict]:
        """
        Get all outgoing edges from a node.

        Args:
            node_id: The node's id property value

        Returns:
            List of dicts with 'source', 'target', 'r' keys
        """
        result = self._conn.cypher(
            "MATCH (a {id: $id})-[r]->(b) "
            "RETURN a.id AS source, b.id AS target, r",
            params={"id": node_id},
        )
        return result.to_list()

    def get_edges_to(self, node_id: str) -> list[dict]:
        """
        Get all incoming edges to a node.

        Args:
            node_id: The node's id property value

        Returns:
            List of dicts with 'source', 'target', 'r' keys
        """
        result = self._conn.cypher(
            "MATCH (a)-[r]->(b {id: $id}) "
            "RETURN a.id AS source, b.id AS target, r",
            params={"id": node_id},
        )
        return result.to_list()

    def get_edges_by_type(self, node_id: str, rel_type: str) -> list[dict]:
        """
        Get outgoing edges of a specific type from a node.

        Args:
            node_id: The node's id property value
            rel_type: The relationship type to filter by

        Returns:
            List of dicts with 'source', 'target', 'r' keys
        """
        from ..utils import sanitize_rel_type
        safe_type = sanitize_rel_type(rel_type)
        result = self._conn.cypher(
            f"MATCH (a {{id: $id}})-[r:{safe_type}]->(b) "
            f"RETURN a.id AS source, b.id AS target, r",
            params={"id": node_id},
        )
        return result.to_list()

    def stats(self) -> dict[str, int]:
        """
        Get graph statistics.

        Returns:
            Dict with 'nodes' and 'edges' counts
        """
        nodes = self._conn.cypher("MATCH (n) RETURN count(n) AS cnt")
        edges = self._conn.cypher("MATCH ()-[r]->() RETURN count(r) AS cnt")

        node_cnt = nodes[0].get("cnt", 0) if len(nodes) > 0 else 0
        edge_cnt = edges[0].get("cnt", 0) if len(edges) > 0 else 0

        return {
            "node_count": int(node_cnt) if node_cnt else 0,
            "edge_count": int(edge_cnt) if edge_cnt else 0,
        }

    def query(self, cypher: str, params: Optional[dict] = None) -> list[dict]:
        """
        Execute a raw Cypher query with optional parameters.

        Args:
            cypher: Cypher query string (may contain $param placeholders)
            params: Optional dictionary of parameter values

        Returns:
            List of result dictionaries
        """
        result = self._conn.cypher(cypher, params)
        return result.to_list()
